keps/sig-api-machinery/1040-priority-and-fairness/README.md

```diff
@@ -1066,8 +1076,361 @@ the remaining requests in that queue start getting faster service.  In
+a request, we should first estimate how expensive a particular request is.
+And we need to think about both dimensions that we're trying to protect from
+overloading (CPU and RAM) and how many concurrency units a request can actually
+consume.
+
+Let's start with CPU.  The total cost of processing a LIST request should be
+proportional to the number of processed objects.  However, given that in
+practice processing a single request isn't parallelized (and the fact that
+we generally scale the number of total concurrency units linearly with amount
+of available resources), a single request should consume no more than A
+concurrency units.  Fortunately that all compiles together because the
+`processing latency` of the LIST request is actually proportional to the
+number of processed objects, so the cost of the request (defined above as
+`<width> x <processing latency>` really is proportaional to the number of
+processed objects as expected.
+
+For RAM the situation is actually different.  In order to process a LIST
+request we actually store all objects that we process in memory. Given that
+memory is uncompressable resource, we effectively need to reserve all that
+memory for the whole time of processing that request.  That suggests that
+the `width` for the request from the RAM perspective should be proportional
+to the number of processed items.
+
+So what we get is that:
+```
+  width_cpu(N) = min(A, B * N)
+  width_ram(N) = C * N
+```
+where N is the number of items a given LIST request is processing.
+
+The question is how to combine them to a single number.  While the main goal
+is to stay on the safe side and protect from the overload, we also want to
+maxiumize the utilization of the available concurrency units.
+Fortunately, when we normalize CPU and RAM to percentage of available capacity,
+it appears that almost all requests are much more cpu-intensive.  Assuming
+4GB:1CPU ratio and 10kB average object and the fact that processing larger
+number of objects can utilize exactly 1 core, that means that we need to
+process 400.000 objects to make the memory cost higher.
+This means, that we can afford the potential minor efficiency that extremely
+large requests would cause and just approximate it by protecting every resource
+independently, which translates to the following function:
+```
+  width(n) = max(min(A, B * N), C * N)
+```
+We're going to better tune the function based on experiments, but we will
+simply start with
+```
+width_approx(n) = max(A, ceil(N / B))
+```
+Fortunately that logic will be well separated and purely in-memory so we
+can decide to arbitrarily adjust it in future releases.
+
+Given that the estimation is well separated piece of logic, we can decide
+to replace with much more sophisticated logic later (e.g. whether it is
+served from etcd or from cache, whether it is namespaced or not, etc.).
+
+One more important aspect to resolve is what happens if a given priority
```

Comment:
I think the questions that generalizing width poses for dispatching are as follows.

Fairness.  Obviously width contributes to `rho`.  Fairness comes from the allocated `mu` values, which are not challenged by generalizing width.

What happens in the real world?  Baseline answer is: a dispatch can happen when the _available_ concurrency is at least the width of the next request to dispatch.  What happens if that width is larger than the total available concurrency?  I think that in this case we have make an except and allow dispatch of such a big request when nothing else is running.

What happens in the virtual world?  When every width is 1, a queue runs `min(rho, C)` requests at once.  With general width, this needs more careful definition.  We could say:
1. a queue runs the shortest prefix of its queue that consumes at least `min(rho, C)`; or
2. a queue runs the longest prefix of its queue that consumes at most `min(rho, C)`; or
3. a queue runs the prefix of its queue that consumes exactly `min(rho, C)`, and we define what it means to run a request with less concurrency than we estimate it needs; I think the obvious answer is that it will take proportionally longer.

Remember that we need to make sure that requests do not finish in the virtual world before they finish in the real world.  This rules out answer 1.  Of the remaining answers, I prefer 2 because it is simpler.

Note also that width is estimated up front.  If we do not adjust for actual amount of output then the accuracy of the estimates becomes critical.  When a request finishes in the real world, we can know its output volume and adjust its width.  I do not want to get into the business of adjusting the past.  But I think we can adjust the remainder of the virtual execution of the request.

Reply:
I think (3) is quite difficult to reason about (because as described above we can't infinitely trade width for the length). So I wouldn't even try to go this route.

But I think I don't like neither option (1) nor (2) either, because in both cases they can actually result in making progress faster than in real world (which sounds problematic for me).


Actually thinking more about what you said during the meeting, I realized that I may not understand something.
I think that the current implementation works as the requests that are currently in-flight are exactly the ones that are in-progress in the virtual world (we push the virtual time of the queue when requests starts by its estimated cost, and adjust back when it finishes). Do I misunderstand something?
If so - what this KEP is proposing is basically same, which is - in virtual world the requests that are processed are exactly the same that are being processed in real world.

Reply:
It's a 2d bin packing problem, but I think our estimates of both width and duration are going to be inaccurate enough that we don't need to focus on scheduling.

The most obvious selection policy to me is to keep the existing round-robin behavior and accept requests as long as the sum of currently executing widths is less than the available width/seats--this way we don't have to figure out what to do with single requests that are too wide. I am totally certain we can come up with an improvement on this, but it's very easy to implement (I think?) and should give a good baseline for further improvements.

Reply:
@wojtek-t : Remember that the implementation lags behind the KEP.  Before this here KEP merged, the KEP was already ahead of the implementation in an important way: the KEP described a way to achieve max-min fairness, while this implementation is older and aims for an even division.

My approach 1 clearly does not risk executing faster in the virtual world than the real world.  It is exactly the other way around.  By spreading the capacity over more requests than would happen in the real world, the virtual completions are not going to happen before the real ones.  For example, spreading a capacity of 600 over 1000 virtual concurrent requests means that each one progresses at a rate of 0.6 virtual work-seconds per real second.

My approach 2 also does not risk completing early in the virtual world, because width _does not_ trade off with length --- at least in the virtual world.  Unused capacity in the virtual world does not speed up the virtually executing requests.

@lavalamp : we _do_ have to worry about the virtual scheduling and the real scheduling.  I am not sure I understand your words as you intend.  I do not see how you addressed with problem that happens when a request's width is greater than a priority level's concurrency limit.

Reply:
@MikeSpreitzer 

> My approach 1 clearly does not risk executing faster in the virtual world than the real world. It is exactly the other way around. By spreading the capacity over more requests than would happen in the real world, the virtual completions are not going to happen before the real ones. For example, spreading a capacity of 600 over 1000 virtual concurrent requests means that each one progresses at a rate of 0.6 virtual work-seconds per real second.

I guess I might have misunderstood your intention here. So let me rephrase to check if I understand i now correctly.
What you're suggesting is that:
- in a *virtual world", the queue is runs the shortest prefix that consumes at least min(rho, C)
- but in real world, we don't dispatch that until we actually have enough seats
Is that what you have on your mind?


>  I do not see how you addressed with problem that happens when a request's width is greater than a priority level's concurrency limit.

See https://github.com/kubernetes/enhancements/pull/2632/files#diff-0d81f61bda1a026d678a311aa592601d41014434a5a3c91e17a3a3dfe1d4527aR1179-R1201